src/Interpreters/RewriteArrayExistsFunctionVisitor.cpp

```diff
@@ -76,4 +77,22 @@ void RewriteArrayExistsFunctionMatcher::visit(const ASTFunction & func, ASTPtr &
     }
 }
 
+bool RewriteArrayExistsFunctionMatcher::needChildVisit(const ASTPtr & ast, const ASTPtr & child)
+{
+    /// SELECT ... JOIN ... ON arrayExists(x -> x = 1, arr)
```

Comment:
The comment is not explanatory enough for such a specific code (why only `arrayExists`? smells badly)

Reply:
Changed the implementation to correctly apply the optimization even for incorrect JOIN ON expression.